pandas/core/generic.py

```diff
@@ -4110,6 +4077,133 @@ def ranker(data):
+              fractional part of the index surrounded by `i` and `j`.
+            * lower: `i`.
+            * higher: `j`.
+            * nearest: `i` or `j` whichever is nearest.
+            * midpoint: (`i` + `j`) / 2.
+
+        Returns
+        -------
+        %s
+
+        Examples
+        --------
+
+        >>> s = Series([1, 2, 3, 4])
+        >>> s.quantile(.5)
+            2.5
+        >>> s.quantile([.25, .5, .75])
+        0.25    1.75
+        0.50    2.50
+        0.75    3.25
+        dtype: float64
+        >>> df = DataFrame(np.array([[1, 1], [2, 10], [3, 100], [4, 100]]),
+                           columns=['a', 'b'])
+        >>> df.quantile(.1)
+        a    1.3
+        b    3.7
+        dtype: float64
+        >>> df.quantile([.1, .5])
+               a     b
+        0.1  1.3   3.7
+        0.5  2.5  55.0
+        """)
+
+    @Appender(_shared_docs['quantile'] % '')
+    def quantile(self, q=0.5, axis=0, numeric_only=None,
+                 interpolation='linear'):
+        if self.ndim >= 3:
+            msg = "quantile is not implemented on on Panel or PanelND objects."
+            raise NotImplementedError(msg)
+        elif self.ndim == 1:
+            result = self.to_frame().quantile(q=q, axis=axis,
+                                              numeric_only=numeric_only,
+                                              interpolation=interpolation)
+            if not com.is_list_like(q):
+                return result.iloc[0]
+            else:
+                return result[result.columns[0]]
+
+        self._check_percentile(q)
+        per = np.asarray(q) * 100
+
+        if not com.is_list_like(per):
+            per = [per]
+            q = [q]
+            squeeze = True
+        else:
+            squeeze = False
+
+        if _np_version_under1p9:
+            if interpolation != 'linear':
+                raise ValueError("Interpolation methods other than linear "
+                                 "are not supported in numpy < 1.9")
+
+        def f(arr, per, interpolation):
+            if arr._is_datelike_mixed_type:
+                values = _values_from_object(arr).view('i8')
+            else:
+                values = arr.astype(float)
+            values = values[notnull(values)]
+            if len(values) == 0:
+                return NA
+            else:
+                if _np_version_under1p9:
+                    return _quantile(values, per)
+                else:
+                    return _quantile(values, per, interpolation=interpolation)
+
+        data = self._get_numeric_data() if numeric_only else self
+
+        axis = self._get_axis_number(axis)
+
+        if axis == 1:
+            data = data.T
+
+        # need to know which cols are timestamp going in so that we can
+        # map timestamp over them after getting the quantile.
+        is_dt_col = data.dtypes.map(com.is_datetime64_dtype)
+        is_dt_col = is_dt_col[is_dt_col].index
```

Comment:
do all of this stuff in ``f`` itself, it should return a Series that is of the correct dtype.